logstage/logstage-api-logger/src/main/scala/com/github/pshirshov/izumi/logstage/api/rendering/StringRenderingPolicy.scala

```diff
@@ -17,145 +20,59 @@ class StringRenderingPolicy(options: RenderingOptions) extends RenderingPolicy {
+              }
           }
-        case None =>
       }
     }
 
-    builder.toString()
-  }
-
-  protected def formatKv(kv: (String, Any)): String = {
-    if (withColors) {
-      s"${Console.GREEN}${kv._1}${Console.RESET}=${Console.CYAN}${kv._2}${Console.RESET}"
-    } else {
-      s"${kv._1}=${kv._2}"
-    }
-  }
-
-  def formatMessage(entry: Log.Entry): RenderedMessage = {
-    val templateBuilder = new StringBuilder()
-    val messageBuilder = new StringBuilder()
-    //    val rawMessageBuilder = new StringBuilder()
-
-    val head = entry.message.template.parts.head
-    templateBuilder.append(StringContext.treatEscapes(head))
-    messageBuilder.append(StringContext.treatEscapes(head))
-    //    rawMessageBuilder.append(head)
-
-    val balanced = entry.message.template.parts.tail.zip(entry.message.args)
-    val unbalanced = entry.message.args.takeRight(entry.message.args.length - balanced.length)
-    balanced.foreach {
-      case (part, (argName, argValue)) =>
-        templateBuilder.append('{')
-        templateBuilder.append(argName)
-        templateBuilder.append('}')
-        templateBuilder.append(StringContext.treatEscapes(part))
-
-        messageBuilder.append(formatKv((argName, argToString(argValue))))
-        messageBuilder.append(StringContext.treatEscapes(part))
-
-      //        rawMessageBuilder.append('{')
-      //        rawMessageBuilder.append(argName)
-      //        rawMessageBuilder.append('=')
-      //        rawMessageBuilder.append(argToString(argValue))
-      //        rawMessageBuilder.append('}')
-
-    }
-
-    unbalanced.foreach {
-      case (argName, argValue) =>
-        templateBuilder.append("; ?")
-        messageBuilder.append("; ")
-        messageBuilder.append(formatKv((argName, argToString(argValue))))
+    logUnitFinder.appendTail(sb)
+    if (options.withExceptions) {
+      sb.append(traceThrowable(entry))
     }
+    sb.toString
 
-    RenderedMessage(entry, templateBuilder.toString(), messageBuilder.toString())
   }
 
 
-  protected def argToString(argValue: Any): String = {
-    argValue match {
-      case e: Throwable =>
+  def traceThrowable(entry: Log.Entry): String = {
+    entry.firstThrowable match {
+      case Some(t) =>
+        val builder = new StringBuilder
+        builder.append('\n')
         if (withColors) {
-          s"${Console.YELLOW}${e.toString}${Console.RESET}"
-        } else {
-          e.toString
+          builder.append(Console.YELLOW)
         }
-
-      case _ =>
-        argValue.toString
+        import IzThrowable._
+        builder.append(t.stackTrace)
+        if (withColors) {
+          builder.append(Console.RESET)
+        }
+        builder.toString()
+      case None =>
+        ""
     }
   }
-
 }
 
 
+object StringRenderingPolicy {
+  val logUnitMatcher: String => Matcher = Pattern.compile("(\\$\\{[^}]+\\})").matcher(_: String)
```

Comment:
replaced regex at all. processing of requested policy layout goes on StringRenderingPolicy initialization only once.